### ISSUE TITLE
Websocket Handler Race Condition

### ISSUE THREAD

**Issue by nyoung-figly** [ztellman/aleph#242]
(I think)

We've been seeing a sporadic issue with our server dropping the first message sent by clients after connecting via websocket. From adding logging everywhere we possibly could, it looks to me like the `initialize-websocket-handler` performs the handshake and then wraps that result, a netty `ChannelFuture`, with a manifold deferred.

The upshot here is that the netty event pool is still processing read events from the client and the manifold pool is handling the installation the websocket netty handler into the pipeline, resulting in a race to see what happens first: either the client sends the message off and it gets to the end of the netty pipeline where its discarded, or the manifold deferred executes its success handler first, installing the websocket handler, and the message is properly passed along to our code.

Here's a sample from a log file in which we saw the problem occur. You can see the "Discarded inbound message" occur before the websocket-handler (a log statement we inserted into aleph) is added:

```
12464 [aleph-netty-server-event-pool-1-2] DEBUG io.netty.buffer.AbstractByteBuf  - -Dio.netty.buffer.bytebuf.checkAccessible: true
12471 [aleph-netty-server-event-pool-1-2] DEBUG io.netty.util.Recycler  - -Dio.netty.recycler.maxCapacity.maxCapacity: 262144
12572 [manifold-pool-2-1] DEBUG io.netty.handler.codec.http.websocketx.WebSocketServerHandshaker  - [id: 0x359d772e, L:/172.19.0.3:3001 - R:/192.168.99.1:59086] WebSocket version V13 server handshake
12593 [manifold-pool-2-1] DEBUG io.netty.util.internal.Cleaner0  - java.nio.ByteBuffer.cleaner(): available
12594 [manifold-pool-2-1] DEBUG io.netty.handler.codec.http.websocketx.WebSocketServerHandshaker  - WebSocket version 13 server handshake key: TcufyL1pepjVoWYssBCb8w==, response: Xz6IZQ114BxyExfFX4vZk3JKcwc=
12627 [manifold-pool-2-1] DEBUG io.netty.util.internal.JavassistTypeParameterMatcherGenerator  - Generated: io.netty.util.internal.__matchers__.io.netty.handler.codec.http.websocketx.WebSocketFrameMatcher
12661 [manifold-pool-2-1] DEBUG aleph.http.server  - removed request-handler
12707 [aleph-netty-server-event-pool-1-2] DEBUG io.netty.handler.codec.http.websocketx.WebSocket08FrameDecoder  - Decoding WebSocket Frame opCode=1
12708 [aleph-netty-server-event-pool-1-2] DEBUG io.netty.handler.codec.http.websocketx.WebSocket08FrameDecoder  - Decoding WebSocket Frame length=87
12709 [aleph-netty-server-event-pool-1-2] DEBUG io.netty.channel.DefaultChannelPipeline  - Discarded inbound message TextWebSocketFrame(data: PooledUnsafeDirectByteBuf(ridx: 0, widx: 87, cap: 87)) that reached at the tail of the pipeline. Please check your pipeline configuration.
12726 [manifold-pool-2-1] DEBUG aleph.http.server  - added websocket-frame-aggregator
12726 [manifold-pool-2-1] DEBUG aleph.http.server  - added websocket-handler
```

**Comment by ztellman** [comment]
Hmmm.  What *should* happen here is that the handshake should complete and the handler get inserted as an atomic action, so there's no period of time where a message should come in unless the client is sending messages before the handshake is complete.  I'll take a look, though.

**Comment by nyoung-figly** [comment]
I threw a `(Thread/sleep 5000)` here: https://github.com/ztellman/aleph/blob/master/src/aleph/http/server.clj#L537 just before the `.remove` and now every message I send to the server for 5s after connection gets dropped and every one after that gets received.

That suggests to me that the handler insertion is atomic, but it is run concurrently with read events and event-handlers. I think you want some way to block reads from being processed by the pipeline after the handshake is complete and the websocket handler can be installed. Not sure I know enough netty to say if that's possible, though.

**Comment by ztellman** [comment]
In Netty, everything for a particular connection runs on a single thread, so calling `Thread/sleep` should block the reads along with everything else.  I believe the `deferred` may be throwing it onto the Aleph thread pool, however.  I'll confirm, and if so it's an easy fix.  Thanks for reporting this.

**Comment by ztellman** [comment]
This should be fixed in the latest release, `0.4.2-alpha2`.
